cpp/src/arrow/csv/reader.cc

```diff
@@ -391,17 +391,20 @@ namespace {
 class BlockParsingOperator {
  public:
   BlockParsingOperator(io::IOContext io_context, ParseOptions parse_options,
-                       int num_csv_cols, int64_t first_row)
+                       int num_csv_cols, int64_t first_row, int64_t offset)
       : io_context_(io_context),
         parse_options_(parse_options),
         num_csv_cols_(num_csv_cols),
         count_rows_(first_row >= 0),
-        num_rows_seen_(first_row) {}
+        num_rows_seen_(first_row),
+        next_block_offset_(offset) {}
 
   Result<ParsedBlock> operator()(const CSVBlock& block) {
     constexpr int32_t max_num_rows = std::numeric_limits<int32_t>::max();
-    auto parser = std::make_shared<BlockParser>(
-        io_context_.pool(), parse_options_, num_csv_cols_, num_rows_seen_, max_num_rows);
+    next_block_offset_ += block.bytes_skipped;
```

Comment:
At the moment streaming parsing is always serial but that is due to a bug/limitation (ARROW-13155).  Eventually, `operator()` will be called multiple times at once (across multiple blocks).  I think this breaks both `num_rows_seen_` and `next_block_offset_` but we can fix it then.